### PR TITLE
C7n mailer ldap cleanup

### PR DESCRIPTION
**Pull request by JohnTheodore**
This was written in a way so you have to opt-in to using LDAP explicitly, think of it as the tag is hard typed, versus dynamically typed. One of the reasons the c7n_mailer code is so confusing and tricky to work on, is because of the polymorphic input parameters, eg sns topic, ldap username, or email, or some other random string. I think if we separated out these values into different target types/tags, the code would get a lot simpler/easier to read. For ldap's sake, I think it needs to be it's own dedicated tag, so we don't spam the AD instance with random strings.

What you get with this PR:
- optional (default off) multiprocessing support for c7n_mailer (called programatically, not from CLI)
- you can configure your template file as an absolute path anywhere on the filesystem (you're not locked into the c7n_mailer/msg-templates directory)
- removing old code that wasn't being run/general cleanup
- a lot less log spam (and more condensed). You can still get all the original logs if you set it for debug mode (which can be set from the start). by default it just does log.info for each message that is sent and tells you what account, what policy, what template, and who (email or SNS) it was sent to, for how many qty of resource for each message (a lot less stuff going to output).
- the ldap mailer code is now usable by other people. it's explicitly it's own target type, like event-owner or resource-owner. Since we don't want to spam folks AD instances with all these random strings. you have to explicitly declare a tag as an ldap username, and then we'll query LDAP for it (still with all the cache support).
- you can now specify to email the manager of the resource owner too (it looks like someone tried to set that up, but it wasn't working afaict, I cleaned it up and got it working)
- you now explicitly set html or text as the format of your email in the action -> notify section, it default to html. this was we don't rely on the file extension, which isn't that reliable and confuses people.
- you can set an option in the mailer.yml to not do the kms base64 decode/decrypt stuff (though by default it's on)

### CODE REVIEW

**Comment by kapilt** [comment]
can you merge trunk and push again, drone server had a memory issue, and missed picking this up, also minor conflict on the readme. thanks.

File: tools/c7n_mailer/README.md [hunk 1]
```diff
@@ -164,6 +169,7 @@ policies:
     actions:
       - type: notify
         template: default
+        template_format: 'html'
```
Comment (JohnTheodore) [inline tools/c7n_mailer/README.md]: Explicitly set the 'html' or 'text' format in the notify action section (it defaults to html).

File: tools/c7n_mailer/c7n_mailer/cli.py [hunk 2]
```diff
@@ -35,6 +35,10 @@
         'ldap_uri': {'type': 'string'},
         'ldap_bind_dn': {'type': 'string'},
         'ldap_bind_user': {'type': 'string'},
+        'ldap_username_attribute': {'type': 'string'},
+        'ldap_manager_attribute': {'type': 'string'},
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/cli.py]: If folks have different attributes, like uid instead of samaccountname they can set those for their specific AD schema

File: tools/c7n_mailer/c7n_mailer/handle.py [hunk 3]
```diff
@@ -23,12 +23,13 @@
...
 
 
-def config_setup(session):
+def config_setup(session, config=None):
     task_dir = os.environ.get('LAMBDA_TASK_ROOT')
     os.environ['PYTHONPATH'] = "%s:%s" % (task_dir, os.environ.get('PYTHONPATH', ''))
-    with open(os.path.join(task_dir, 'config.json')) as fh:
-        config = json.load(fh)
-    if config['ldap_bind_password']:
+    if not config:
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/handle.py]: keep the capone code working for KMS decrypting the LDAP password, but allow an option to not do this. perhaps this should default to False instead of True, thoughts?

File: tools/c7n_mailer/c7n_mailer/handle.py [hunk 4]
```diff
@@ -40,11 +41,13 @@ def config_setup(session):
     return config
 
 
-def start_c7n_mailer(event, context, logger):
+def start_c7n_mailer(event, context, logger, config=None, parallel=False):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/handle.py]: allow someone who is calling the mailer directly as a function (instead of lambda) to decide to use parallelization, parallel is False by default, so it will run in the normal threaded mode unless you decide to do multiprocessing explicitly.

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+# limitations under the License.
+import logging
+import redis
+import re
+import shelve
+from ldap3 import Connection
+from ldap3.core.exceptions import LDAPSocketOpenError
+
+
+class LdapLookup(object):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: turn the ldap code into it's own class/object

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            )
+        self.log          = logger
+        self.base_dn      = config.get('ldap_bind_dn')
+        self.email_key    = config.get('ldap_email_key', 'mail')
+        manager_attr      = config.get('ldap_manager_attribute', 'manager')
+        self.attributes   = ['displayName', self.email_key, manager_attr]
+        self.uid_key      = config.get('ldap_username_attribute', 'sAMAccountName')
+        self.uid_regex    = config.get('ldap_uid_regex', None)
+        self.cache_engine = config.get('cache_engine', None)
+        if self.cache_engine == 'redis':
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: switch to using redis, this has the same get and set methods as memcache

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            redis_host = config.get('redis_host')
+            redis_port = int(config.get('redis_port', 6379))
+            self.caching = redis.StrictRedis(host=redis_host, port=redis_port, db=0)
+        elif self.cache_engine == 'localshelf':
+            self.caching = LocalShelf('/tmp/ldap.cache')
+
+    def get_connection(self, ldap_uri, ldap_bind_user, ldap_bind_password):
+        # note, if ldap_bind_user and ldap_bind_password are None
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: note, now you can anonymous bind, previously it required creds or it failed.

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            conn = None
+        return conn
+
+    def get_email_from_metadata(self, ldap_user_metadata):
+        return ldap_user_metadata.get(self.email_key, None)
+
+    def user_metadata_has_manager(self, ldap_user_metadata):
+        return ldap_user_metadata.get('manager', None)
+
+    def get_metadata_from_dn(self, user_dn):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: Since the 'manager' default AD schema has the manager as a DN, or direct path to the tree on AD, we need a separate instance method to do looks from a DN, versus doing lookups from a UID.

Since these are separate methods for doing lookups, we also store the keys different. There are key/value pairs for DNs, and different key / same values for UIDs

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            if cache_result:
+                return cache_result
+        ldap_filter = '(%s=*)' % self.uid_key
+        self.connection.search(user_dn, ldap_filter, attributes=self.attributes)
+        ldap_user_metadata = self.get_dict_from_ldap_object(self.connection.entries[0])
+        if self.cache_engine:
+            self.caching.set(user_dn, ldap_user_metadata)
+        return ldap_user_metadata
+
+    def get_toaddr_from_ldap_metadata(self, ldap_user_to_email_map, ldap_username, email_manager):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: a helper method to let you get the user email and the manager email in an array read to be set as the 'to' field in the Mimetext message. This is a convenient helper method for the SqsMessageProcessor object to use.

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        return ldap_user_metadata
+
+    def get_metadata_from_uid(self, uid):
+        if self.uid_regex:
+            # for example if you set ldap_uid_regex in your mailer.yml to "^[0-9]{6}$" then it
+            # would only query LDAP if your string length is 6 characters long and only digits.
+            # re.search("^[0-9]{6}$", "123456")
+            # Out[41]: <_sre.SRE_Match at 0x1109ab440>
+            # re.search("^[0-9]{6}$", "1234567") returns None, or "12345a' also returns None
+            if not re.search(self.uid_regex, uid):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: note, this regex makes it easy to filter away potential queries to AD. So you can require (as an example) that your UID is 6 characters long always, and includes numbers only. ( "^[0-9]{6}$" )

Or you can require an underscore to be in the UID, each unique schema and how they setup their UID can be accommodated to reduce the load on AD.

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        if self.cache_engine:
+            self.caching.set(uid, ldap_user_metadata)
+        return ldap_user_metadata
+
+
+# the purpose of this class is to create a light wrapper around a standard library called shelve
+# which can store key/value pairs to a local file (using pickle), similar to how memcache or redis
+# stores things. This normalizes the methods to set/get a function, so you can interchangeable
+# decide which caching system to use, a local file, or memcache, redis, etc
+class LocalShelf(object):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: If someone doesn't want to deal with redis as a dependency, they can just use a local file. This shouldn't be used in lambda, but works great say if you're calling the mailer directly from your container or EC2 instance. all the methods in this object are normalized so they're compatible with redis or memcache.

File: tools/c7n_mailer/c7n_mailer/record.py [hunk 12]
```diff
@@ -1,150 +0,0 @@
-# Copyright 2016 Capital One Services, LLC
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/record.py]: This whole file was thrown into utils, there seem to be various random functions in here, might as well condense it into utils.

File: tools/c7n_mailer/c7n_mailer/record.py [hunk 13]
```diff
@@ -1,150 +0,0 @@
...
-
-
-def resource_tag(resource, k):
-    for t in resource.get('Tags', []):
-        if t['Key'] == k:
-            return t['Value']
-    return ''
-
-
-def resource_owner(resource):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/record.py]: This is old code that doesn't seem to have been used at all, it was removed.

File: tools/c7n_mailer/c7n_mailer/record.py [hunk 14]
```diff
@@ -1,150 +0,0 @@
...
-    return parser.parse(utc_str).astimezone(gettz(tz_str)).strftime(format)
-
-
-def format_struct(evt):
-    io = StringIO()
-    json.dump(evt, io, indent=2)
-    return io.getvalue()
-
-
-def resource_tag(resource, k):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/record.py]: this also doesn't look like it's being used at all, removed

File: tools/c7n_mailer/c7n_mailer/replay.py [hunk 15]
```diff
@@ -1,178 +0,0 @@
-"""
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/replay.py]: this whole file doesn't seem hooked up to anything? or working.. removed

File: tools/c7n_mailer/c7n_mailer/sqs_queue_processor.py [hunk 16]
```diff
@@ -16,7 +16,7 @@
 ===============
 
 """
-import memcache
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/sqs_queue_processor.py]: move all the caching logic directly into the LdapLookup class, since that's the only class that uses the logic it probably belongs there

File: tools/c7n_mailer/c7n_mailer/sqs_queue_processor.py [hunk 17]
```diff
@@ -29,7 +29,7 @@ class MailerSqsQueueIterator(object):
     def __init__(self, aws_sqs, queue_url, logger, limit=0, timeout=10):
         self.aws_sqs   = aws_sqs
         self.queue_url = queue_url
-        self.limit     = limit or limit
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/sqs_queue_processor.py]: not sure why this was like that, simplified

File: tools/c7n_mailer/c7n_mailer/utils.py [hunk 18]
```diff
@@ -11,14 +11,13 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from cStringIO import StringIO
-import json
 
+import datetime
+import json
 
-def format_struct(struct):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/utils.py]: This is a duplicate of the same function in another file, but this one wasn't used. removed

File: tools/c7n_mailer/c7n_mailer/sqs_message_processor.py [hunk 19]
```diff
@@ -20,12 +20,11 @@
 import yaml
 import zlib
 
-from address import get_user
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/sqs_message_processor.py]: 'address' is a bit confusing, since there are so many address, sns, email, etc. made the file name more descriptive

File: tools/c7n_mailer/c7n_mailer/sqs_message_processor.py [hunk 20]
```diff
@@ -20,12 +20,11 @@
...
+from ldap_lookup import LdapLookup
 from boto3 import Session
 from email.mime.text import MIMEText
 from email.utils import parseaddr
-from functools import partial
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/sqs_message_processor.py]: no need to do partials anymore, we just have a long lived LdapLookup object that is an instance variable for each mailer run.

File: tools/c7n_mailer/c7n_mailer/sqs_message_processor.py [hunk 21]
```diff
@@ -40,37 +39,33 @@ class SqsMessageProcessor(object):
...
+        self.ldap_lookup = LdapLookup(config, logger)
 
-        self.env.loader = jinja2.FileSystemLoader(
+        self.env.loader = jinja2.FileSystemLoader([
             os.path.abspath(
                 os.path.join(
                     os.path.dirname(os.path.abspath(__file__)),
                     '..',
-                    'msg-templates')))
+                    'msg-templates')), os.path.abspath('/')])
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/sqs_message_processor.py]: allow someone to specify the absolute directory of their jinja template file. This way you're not locked into the msg-templates directory in c7n_mailer. You can just set 'template: /tmp/foo.jinja' as an example.

File: tools/c7n_mailer/c7n_mailer/sqs_message_processor.py [hunk 22]
```diff
@@ -107,67 +102,118 @@ def process_sqs_messsage(self, message):
...
+            targets.remove('ldap_username_tags')
+            self.send_ldap_username_tags_email(data)
         if not targets:
-            self.logger.info("No intended targets found for message")
+            self.logger.debug("No more intended targets found for message")
             return
-        self.send_message_to_targets(targets, data, data['resources'])
+        self.send_message_to_targets(targets, data)
+
+    def send_ldap_username_tags_email(self, data):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/sqs_message_processor.py]: this is the main function that does an email lookup in LDAP and sends off emails. note you can add multiple tags that you expect to be LDAP usernames. though most folks will likely only use one tag (the auto-tagger tag they use for aws_username)

File: tools/c7n_mailer/c7n_mailer/sqs_message_processor.py [hunk 23]
```diff
@@ -178,48 +224,48 @@ def send_message_to_targets(self, targets, data, resources):
                 continue
             self.send_c7n_email(data, email_to_addr, subject, body)
 
-    def send_c7n_email(self, data, email_to_addr, subject, body):
-        email_format = data['action'].get(
-            'template', 'default').endswith('html') and 'Html' or 'Text'
+    def send_smtp_email(self, smtp_server, message):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/sqs_message_processor.py]: The send_c7n_email function was a bit bit, we we broken off the smtp part into it's own dedicated function.

File: tools/c7n_mailer/c7n_mailer/sqs_message_processor.py [hunk 24]
```diff
@@ -273,59 +320,27 @@ def render(self, target, data, resources):
             action=data['action'],
             policy=data['policy'],
             region=data.get('region', ''))
-
         return message
 
-    def get_policy_targets(self, data):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/sqs_message_processor.py]: this is another function that doesn't seem to be called anywhere, it's removed.

**Comment by JohnTheodore** [comment]
https://github.com/capitalone/cloud-custodian/issues/706

**Comment by JohnTheodore** [comment]
You can set 'debug: true' in mailer.yml and have all logging information dumped to output, by default it dumps interesting things, like "INFO - Sending account:core-services-dev policy:ebs-mark-unattached-deletion ebs:1 email:/custodian/email/jinja_template.j2 to [u'bob@example.com', 'bobs_boss@example.com']"

added a regex variable to the mailer called ldap_uid_regex, you can set this to say: "^[0-9]{6}$" and if your uid lookups aren't exactly 6 characters and only digits, it will return None. If they are 6 characters long and only digits, it will hit your cache engine. This gives folks some flexibility in reducing the spam and requests that go to their AD servers.

added redis support

added a local file cache (using a stdlib in python 2.7, so no new dependencies for the lambda mailer). if someone wants avoid the redis dependency, they can just store the cache in a local file. the cache engine for a local file, for redis or memcache are normalized so they all have the same get/set methods.

**Comment by JohnTheodore** [comment]
note, it's basically 3 extra lines to keep memcache support (it's not in there currently). it's redis and a local key/value caching system. I know lambda gets packaged and zipped, so I figured having the extra pypi dependency wouldn't be desirable? let me know your thoughts

File: tools/c7n_mailer/c7n_mailer/replay.py [hunk 25]
```diff
@@ -1,178 +0,0 @@
...
-Allow local testing of mailer and templates by replaying an SQS message.
-
```
Comment (kapilt) [inline tools/c7n_mailer/c7n_mailer/replay.py]: this shouldn't have been removed.

File: tools/c7n_mailer/c7n_mailer/handle.py (hunk 3 above)
Comment (kapilt) [inline tools/c7n_mailer/c7n_mailer/handle.py]: lambda code is not secure, anyone in the account can download, no real interest in supporting passwords in plain text.

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py (hunk 6 above)
Comment (kapilt) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: sounds good

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py (hunk 11 above)
Comment (kapilt) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: sounds good

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+                auto_referrals=False,
+            )
+        except LDAPSocketOpenError:
+            conn = None
+        return conn
+
+    def get_email_from_metadata(self, ldap_user_metadata):
+        return ldap_user_metadata.get(self.email_key, None)
+
+    def user_metadata_has_manager(self, ldap_user_metadata):
```
Comment (kapilt) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: both this and get_email_from are extraneous one line methods.

File: tools/c7n_mailer/c7n_mailer/record.py (hunk 12 above)
Comment (kapilt) [inline tools/c7n_mailer/c7n_mailer/record.py]: there mostly display related wrt to a resource, minus the date utility you added, it would be helpful due to the size of this pr to not do unrelated renames as well. the intent was that the record formatter could go away given there's generic formatting in c7n core now.

File: tools/c7n_mailer/c7n_mailer/handle.py (hunk 3 above)
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/handle.py]: sure, agreed lambda isn't secure, but not everyone runs the mailer in lambda. so KMS shouldn't be a hard dependency. I've deployed custodian with openshift/kubernetes, and I already have a secure workflow using secure volumes that get attached to the container. KMS isn't an option for everyone.

File: tools/c7n_mailer/c7n_mailer/utils.py [hunk 27]
```diff
@@ -33,3 +32,132 @@ def setup_defaults(config):
...
+        return "%s %s %s %s %s %s" % (
+            resource['SubnetId'],
+            resource['VpcId'],
+            resource['AvailabilityZone'],
+            resource['State'],
+            resource['CidrBlock'],
+            resource['AvailableIpAddressCount'])
+    else:
+        print "Unknown resource type", resource_type
+        return "%s" % format_struct(resource)
```
Comment (cmedley) [inline tools/c7n_mailer/c7n_mailer/utils.py]: @JohnTheodore FYI - sending users an email to rotate keys will throw Unknown resource type iam-user

Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/utils.py]: hmm, I'm just moving this function from one file to another. What change are you suggesting in this?

Comment (cmedley) [inline tools/c7n_mailer/c7n_mailer/utils.py]: possibly removing the print line or having an elif for iam-user?

**Comment by JohnTheodore** [comment]
- break off all sns logic into it's own class, so the sqs_message_processor class isn't too big
- added the ability to use the current aws session's credentials instead of requiring cross_account on sns delivery
- tested that redis works, wrote a wrapper class for it to normalize it to localshelf (and memcache)
- add a bunch of tests for checking the email workflow (still more to come later)
- added a bunch of tests for LDAP (with mocking code, etc)
- add CLI to run the mailer manually, also you can run the mailer in debug mode or multiprocessing mode from CLI
- manually tested SES and SMTP still works fine
- manually tested multiple ldap tags works (sends to multiple emails, *2 if you do manager also)
- manually tested that AWS SES still works
- if the ldap query returns no result, save that key to cache with a value of None (so we never hit ldap twice for a bad result)
- chunked the sqs_message_processor file into more unit testable functions with return values. a lot of the function didn't return any value, so things were difficult to unit test that way.
- manually test that emails in the policy notify -> to section get sent, as well as emails from resource-owner tags, as well as ldap lookups send emails
- test local file cache works and wrote unit tests around it
- add json schema changes from mailer.yml to cli.py
- commented the code more so it's descriptive about what is happening and why

still need to write more tests

File: tools/c7n_mailer/c7n_mailer/address.py [hunk 28]
```diff
@@ -1,88 +0,0 @@
-# Copyright 2016 Capital One Services, LLC
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/address.py]: rename this file to ldap_lookup.py since it's a more fitting name

File: tools/c7n_mailer/c7n_mailer/cli.py [hunk 29]
```diff
@@ -47,39 +55,82 @@
...
+    parser.add_argument('--max-num-processes', help=max_num_processes_help_msg)
+    group = parser.add_mutually_exclusive_group(required=True)
+    update_lambda_help_msg = 'packages your c7n_mailer, uploads the zip to aws lambda as a function'
+    group.add_argument('--update-lambda', action='store_true', help=update_lambda_help_msg)
+    run_help_msg = 'run c7n-mailer locally, process sqs messages and send emails or sns messages'
+    group.add_argument('--run', action='store_true', help=run_help_msg)
+    return parser
 
+
+def run_mailer_in_parallel(mailer_config, aws_session, logger, max_num_processes):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/cli.py]: you can run the mailer from CLI now, with --debug options, as well as multiprocessing options

File: tools/c7n_mailer/c7n_mailer/deploy.py [hunk 30]
```diff
@@ -28,11 +28,11 @@
 
 logger = logging.getLogger('custodian.mailer')
 log_format = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
-logging.basicConfig(level=logging.DEBUG, format=log_format)
+logging.basicConfig(level=logging.INFO, format=log_format)
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/deploy.py]: default things to INFO, you can decide to run --debug from CLI, or as well you have the option of debug: true in your mailer.yml

File: tools/c7n_mailer/c7n_mailer/ldap_lookup.py [hunk 31]
```diff
@@ -0,0 +1,174 @@
...
+    def get(self, key):
+        return self.shelve.get(str(key))
+
+    def set(self, key, value):
+        self.shelve[key] = value
+
+
+# redis can't write complex python objects like dictionaries as values (the way memcache can)
+# so we turn our dict into a json string when setting, and json.loads when getting
+class Redis(object):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/ldap_lookup.py]: It turns out redis can't write python dict objects like memcache can, it has to take strings. So this wrapper class is necessary to normalize the gets/gets. I just tested this flow with redis on my mac, works fine.

File: tools/c7n_mailer/c7n_mailer/sns_delivery.py [hunk 32]
```diff
@@ -0,0 +1,136 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from boto3 import Session
+from utils import get_message_subject, get_resource_tag_targets, get_rendered_jinja
+
+
+class SnsDelivery(object):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/sns_delivery.py]: break off Sns related functions into it's own class, the other file was too big for awhile

File: tools/c7n_mailer/c7n_mailer/utils.py [hunk 33]
```diff
@@ -11,14 +11,81 @@
...
+from cStringIO import StringIO
+from dateutil import parser
+from dateutil.tz import gettz
 
 
-def format_struct(struct):
-    io = StringIO()
-    json.dump(struct, io, indent=2)
-    return io.getvalue()
+def get_jinja_env():
```
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/utils.py]: Since Sns is it's own class, we take the couple functions email an sns share (jinja_render, get_subject, and get_resource_tag_targets) and throw them into utils.

File: tools/c7n_mailer/c7n_mailer/utils.py (hunk 27 above)
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/utils.py]: I think @kapilt mentioned this was already implemented in c7n. so this function might not live that much longer in general. I'm not sure where it is in c7n.

File: tools/c7n_mailer/tests/test_email.py [hunk 34]
```diff
@@ -0,0 +1,90 @@
+# Copyright 2016 Capital One Services, LLC
```
Comment (JohnTheodore) [inline tools/c7n_mailer/tests/test_email.py]: add some unit tests around email stuff, still need more.

File: tools/c7n_mailer/tests/test_ldap.py [hunk 35]
```diff
@@ -0,0 +1,96 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+from common import destroy_local_ldap_cache, get_ldap_lookup, PETER, BILL
+
+
+class MailerLdapTest(unittest.TestCase):
```
Comment (JohnTheodore) [inline tools/c7n_mailer/tests/test_ldap.py]: add a bunch of ldap unit tests, ldap is pretty solidly covered. the ldap3 mock section of the library was helpful!

**Comment by JohnTheodore** [comment]
also I have a script I wrote that takes 10s of thousands of LDAP records and populates the cache will all of them, it takes <~30 seconds to run. the c7n_mailer is a lot faster if it (basically never) talks to LDAP when it's processing sqs_messages. Thoughts on adding that as a script to c7n_mailer?

**Comment by JohnTheodore** [comment]
`---------- coverage: platform darwin, python 2.7.13-final-0 ----------
Name                                                   Stmts   Miss  Cover
--------------------------------------------------------------------------
tools/c7n_mailer/c7n_mailer/__init__.py                    0      0   100%
tools/c7n_mailer/c7n_mailer/ldap_lookup.py               117     12    90%
tools/c7n_mailer/c7n_mailer/sns_delivery.py               63     53    16%
tools/c7n_mailer/c7n_mailer/sqs_message_processor.py     145     76    48%
tools/c7n_mailer/c7n_mailer/utils.py                     117     92    21%
tools/c7n_mailer/tests/common.py                          40      1    98%
tools/c7n_mailer/tests/test_email.py                      42      0   100%
tools/c7n_mailer/tests/test_ldap.py                       60      0   100%
tools/c7n_mailer/tests/test_sns.py                         6      2    67%
--------------------------------------------------------------------------
TOTAL                                                    590    236    60%`

**Comment by JohnTheodore** [comment]
I have no new changes planed for this PR (I promise). though it will grow as I add more unit tests, that's all I have left.

**Comment by kapilt** [comment]
the additions of tests alone make this worthwhile, drone tests seem to be failing due to an issue with the fakeredis module that's being masked by a libc err msg.

**Comment by JohnTheodore** [comment]
I fixed the fakeredis test, switched to ubuntu xenial since alpine is non-FHS compliant and breaks fakeredis/ctypes.

I added a bunch more emails tests, and some sns_delivery tests.

**Comment by JohnTheodore** [comment]
I'm passing in drone, and passing in travis. for some reason this PR doesn't show the travis check, which means coveralls doesn't get triggered, but it looks all green from here.

**Comment by kapilt** [comment]
re script, a cache warm script direct from ldap .. per exchange schema? 

noted re ci, I feel like I saw travis awol on a few pull requests in the last 24hrs. the drone changes require a followup pull request to regen drone settings, I'll have a look at merging tomorrow.

File: tools/c7n_mailer/c7n_mailer/cli.py (hunk 29 above)
Comment (mwkaufman) [inline tools/c7n_mailer/c7n_mailer/cli.py]: This removed the option to use -c for the config, e.g. the "c7n-mailer -c mailer.yml" is the example usage and no longer works. Also it now requires either a "--update-lambda" or "--run" to execute, should it default to "--update-lambda" for backwards compatibility?

**Comment by mwkaufman** [comment]
I couldn't get this working in my lab which is just trying to send e-mail via SES. Get the below error in my test policy in I think send_c7n_email.

```
2017-06-05 09:36:00,954 - custodian-mailer - WARNING - Error policy:{u'actions': [{u'to': [u'myemail@gmail.com'], u'template': u'default', u'subject': u'testing the c7n mailer', u'type': u'notify', u'transport': {u'queue': u'https://sqs.us-east-1.amazonaws.com/myaccount/c7n-mailer', u'type': u'sqs'}, u'priority_header': 2}], u'resource': u'sqs', u'name': u'c7n-mailer-test', u'filters': [{u'tag:MailerTest': u'absent'}]} account:mylab sending to:[u'myemail@gmail.com']

 error: object of type 'int' has no len()

 mailer.yml: {'subnets': None, 'queue_url': 'https://sqs.us-east-1.amazonaws.com/myaccount/c7n-mailer', 'ses_region': 'us-east-1', 'from_address': 'myemail@gmail.com', 'ldap_bind_user': None, 'region': 'us-east-1', 'contact_tags': [], 'role': 'arn:aws:iam::myaccount:role/lambda-role', 'timeout': 300, 'memory': 1024, 'ldap_bind_dn': None, 'ldap_uri': None, 'security_groups': None, 'ldap_bind_password': None}
```

File: tools/c7n_mailer/c7n_mailer/cli.py (hunk 29 above)
Comment (JohnTheodore) [inline tools/c7n_mailer/c7n_mailer/cli.py]: @mwkaufman so the mailer previously could only do one thing, that is update lambda. Now it does multiple things, update, run, debug, multiprocessing. Just specifying the config option (with no other args), and causing that to update lambda by default seems like confusing behavior (especially for new users). It seems easy enough to just specify --update-lambda, or --run. mutative actions are potentially dangerous, so explicitly opting in to them seems reasonable. I'll leave it up to @kapilt to decide on this though.

**Comment by JohnTheodore** [comment]
@mwkaufman I believe it's because your priority_header on the notify section of your policy is an integer, and should be a string. I explicitly just set it to a string when assigned, I think that will fix your issue.

**Comment by mwkaufman** [comment]
The last commit fixed my issue. I actually don't use priority_header most of the time, but was just using the test policy from the readme and this fixed it. I was able to validate this batches multiple to e-mail addresses into one e-mail correctly which is a big win. It appeared noisier than I expected, but the ack on line 122 of sqs_queue_processor.py is commented out, causing sent messages not to get deleted from the SQS queue.

**Comment by JohnTheodore** [comment]
oi, good catch. fixed the ack. I added that when testing your priority_header issue.

So if things are noisy, you can turn off debug mode. so no --debug in CLI, or don't set debug: true in mailer.yml

I'm glad the batching of emails is a big win for you! Let me know if your testing comes up with anything else when using this PR, I appreciate the feedback!

**Comment by kapilt** [comment]
there's one merge conflict against trunk

**Comment by JohnTheodore** [comment]
fixed conflict and rebased from upstream

**Comment by JohnTheodore** [comment]
@kapilt Let me know if you want any other changes.

**Comment by kapilt** [comment]
mostly just looking to get the conflicts resolved, and the ci checks passing.

**Comment by JohnTheodore** [comment]
It's because a couple times things merged into the mailer before this was merged. :p.

**Comment by JohnTheodore** [comment]
ok, rebased from upstream, fixed conflicts, and I'm passing everything. (for some reason travis ran the first time I pushed, but not the second?). If you want to manually trigger travis in the capitalone space in travis-ci.org that's fine. I can see it passes on my travis space though.

**Comment by kapilt** [comment]
i retriggered the build in travis and it seems to work fine, but for some reason its not posting back on the webhook https://travis-ci.org/capitalone/cloud-custodian/builds/243005159

**Comment by kapilt** [comment]
wondering if the coveralls on tools is overwriting the coveralls from c7n

File: tools/c7n_mailer/c7n_mailer/utils.py (hunk 27 above)
Comment (kapilt) [inline tools/c7n_mailer/c7n_mailer/utils.py]: its part of the reporting for the csv re default fields per resource type
